include/string_utils.h

```diff
@@ -375,4 +379,24 @@ std::optional<float> parse_prefixed_percentage(const char prefix, const std::str
 // returns value only if succeeded
 std::optional<int> to_int(const std::string& value);
 
+template <typename... Args>
+std::string string_format(const std::string& format, Args... args) noexcept
+{
+	constexpr auto buffer   = nullptr;
+	constexpr auto buf_size = 0;
+	auto required_size = std::snprintf(buffer, buf_size, format.c_str(), args...);
+	if (required_size <= 0) {
+		return {};
+	}
+	// For the terminating zero
+	++required_size;
+
+	auto size = static_cast<size_t>(required_size);
+	std::unique_ptr<char[]> buf(new char[size]);
+	std::snprintf(buf.get(), size, format.c_str(), args...);
+
+	const auto size_minus_terminating_zero = size - 1;
+	return std::string(buf.get(), buf.get() + size_minus_terminating_zero);
+}
```

Comment:
A std::string instead of the unique_ptr and new, and the return value of the second snprintf should be checked (it might have failed or be zero), and the size of the formatted string can be chopped down if needed (adjusted):

```c++
template <typename... Args>
std::string string_format(const std::string_view format, Args&&... args) noexcept
{
	// Perform a non-writing format to determine the size
	const auto required_size = std::snprintf(nullptr,
	                                         0,
	                                         format.data(),
	                                         std::forward<Args>(args)...);
	if (required_size <= 0) {
		return {};
	}

	// snprintf's length parameter specifies the maximum number of
	// characters to be written without the trailing null. However, it still
	// writes the trailing null into the buffer so we need to include that
	// in our allocation.
	std::string result(static_cast<size_t>(required_size + 1), '\0');

	const auto result_size = std::snprintf(result.data(),
	                                       required_size,
	                                       format.data(),
	                                       std::forward<Args>(args)...);
	if (result_size <= 0) {
		return {};
	}
	assert(result_size <= result.size());
	result.resize(static_cast<size_t>(result_size));
	return result;
}
```

Reply:
Sure. TBH, i copied this from somewhere and it worked and the guys explanation how it works seemed okay 😄